memtable/write_buffer_manager.cc

```diff
@@ -0,0 +1,115 @@
+//  LICENSE file in the root directory of this source tree. An additional grant
+//  of patent rights can be found in the PATENTS file in the same directory.
+//  This source code is also licensed under the GPLv2 license found in the
+//  COPYING file in the root directory of this source tree.
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#include "rocksdb/write_buffer_manager.h"
+#include <mutex>
+#include "util/coding.h"
+
+namespace rocksdb {
+#ifndef ROCKSDB_LITE
+namespace {
+const size_t kSizeDummyEntry = 1024 * 1024;
+// The key will be longer than blocks in SST files so it won't conflict.
+const size_t kCacheKeyPrefix = kMaxVarint64Length * 4 + 1;
+}  // namespace
+
+struct WriteBufferManager::CacheRep {
+  std::shared_ptr<Cache> cache_;
+  std::mutex cache_mutex_;
+  std::atomic<size_t> cache_allocated_size_;
+  char cache_key_prefix_[kCacheKeyPrefix];
+  uint64_t num_cache_entries_ = 0;
+  std::vector<Cache::Handle*> dummy_handles_;
+
+  Slice GetCacheKey(char* cache_key, uint64_t id) {
+    memcpy(cache_key, cache_key_prefix_, kCacheKeyPrefix);
+    char* end = EncodeVarint64(cache_key + kCacheKeyPrefix, id);
+    return Slice(cache_key, static_cast<size_t>(end - cache_key));
+  }
+};
+#else
+struct WriteBufferManager::CacheRep {};
+#endif  // ROCKSDB_LITE
+
+WriteBufferManager::WriteBufferManager(size_t _buffer_size,
+                                       std::shared_ptr<Cache> cache)
+    : buffer_size_(_buffer_size),
+      memory_used_(0),
+      mutable_memtable_used_(0),
+      cache_rep_(nullptr) {
+#ifndef ROCKSDB_LITE
+  if (cache) {
+    // Construct the cache key using the pointer to this.
+    cache_rep_.reset(new CacheRep());
+    cache_rep_->cache_ = cache;
+    memset(cache_rep_->cache_key_prefix_, 0, kCacheKeyPrefix);
+    size_t pointer_size = sizeof(const void*);
+    assert(pointer_size <= kCacheKeyPrefix);
+    memcpy(cache_rep_->cache_key_prefix_, static_cast<const void*>(this),
+           pointer_size);
+  }
+#endif  // ROCKSDB_LITE
+}
+
+WriteBufferManager::~WriteBufferManager() {
+#ifndef ROCKSDB_LITE
+  if (cache_rep_) {
+    for (auto* handle : cache_rep_->dummy_handles_) {
+      cache_rep_->cache_->Release(handle, true);
+    }
+  }
+#endif  // ROCKSDB_LITE
+}
+
+// Should only be called from write thread
+void WriteBufferManager::ReserveMemWithCache(size_t mem) {
+#ifndef ROCKSDB_LITE
+  assert(cache_rep_ != nullptr);
+  // Use a mutex to protect various data structures. Can be optimzied to a
+  // lock-free solution if it ends up with a performance bottleneck.
+  std::lock_guard<std::mutex> lock(cache_rep_->cache_mutex_);
+
+  size_t new_mem_used = memory_used_.load(std::memory_order_relaxed) + mem;
+  memory_used_.store(new_mem_used, std::memory_order_relaxed);
```

Comment:
fetch_add() guarantees atomicity so it always locks the bus. So it can be slightly more expensive. Here is is already inside the lock so we don't need this atomicity.